plugin/abolish.vim

```diff
@@ -121,7 +121,7 @@ function! s:snakecase(word)
   let word = substitute(a:word,'::','/','g')
```

Comment:
See this? We're already using `/` in an orthogonal way.

The most I would allow is an undocumented coercion.

Reply:
Oh I had not seen that actually. What exactly is the reasoning for this here? Does this need to be a forward-slash?

Reply:
It's for translating `Class::Name` to a file name. It's a bit Ruby specific and perhaps a bad idea in hindsight, but even if I was convinced to rip it out, I don't think I would be open to adding what I deem to be a _different_ bad idea. I don't think `FooBar/Baz` should change to `foo_bar_baz`.

Reply:
Yes I see.

>  I don't think `FooBar/Baz` should change to `foo_bar_baz`.

I agree, that would indeed be unexpected. I tried to motivate this PR in the related issue, #108, like so:

> This would be very helpful for Python development, where one often encounters almost identical strings in dot-case (for module accessing) and "path-case" (when writing documentation and linking to other modules).

I see that having such language-specific features might be out-of-scope for abolish. I am wondering whether there might be a possibility for some sort of modular approach such that a user could register their own coercions :thinking: 
I tried putting something together real quick by injecting my `pathcase` function into `g:Abolish` before abolish gets loaded. While I did get it to work I am facing 2 issues:
- I cannot reverse `/` to another case because I cannot modify the internal `s:snakecase` implementation
- I cannot use the internal `s:snakecase` implementation because it does not appear to be publicly exposes

FWIW here is the small snippet I patched together:
```vim
nnoremap cR <Plug>(abolish-coerce)
vnoremap zr <Plug>(abolish-coerce)

if !exists("g:Abolish")
  let Abolish = {}
endif
call extend(Abolish, {'Coercions': {}}, 'keep')

function! s:function(name) abort
  return function(substitute(a:name,'^s:',matchstr(expand('<sfile>'), '.*\zs<SNR>\d\+_'),''))
endfunction

function! s:snakecase(word)
  let word = substitute(a:word,'::','/','g')
  let word = substitute(word,'\(\u\+\)\(\u\l\)','\1_\2','g')
  let word = substitute(word,'\(\l\|\d\)\(\u\)','\1_\2','g')
  let word = substitute(word,'[.-/]','_','g')
  let word = tolower(word)
  return word
endfunction

function! s:pathcase(word)
  return substitute(s:snakecase(a:word),'_','/','g')
endfunction

call extend(Abolish, {'pathcase': s:function('s:pathcase')}, "keep")
call extend(Abolish.Coercions, {'/': Abolish.pathcase}, "keep")
```

Would you be open to consider:
- exposing `s:snakecase` in some form or another to enable such extensibility?
- possible adding a hook inside of `s:snakecase` which a user could implement in order to inject custom replacements inside of to enable language-specific coercions such as `:: -> /` (your Ruby case) or `/ -> _` (the case I added as part of this PR)

Reply:
Monkey-patching is pretty straightforward if you use `VimEnter`. Here's an example [in rails.vim](https://github.com/tpope/vim-rails/blob/c3ac2889dc448c5442fa2d1096ac918ea8b38f09/plugin/rails.vim#L152-L183). I would expect everything you describe to be possible without changes to abolish.vim.

> * I cannot reverse `/` to another case because I cannot modify the internal `s:snakecase` implementation

Doesn't sound like you need to. Override the `_` coercion, or whatever.

> * I cannot use the internal `s:snakecase` implementation because it does not appear to be publicly exposes

You can call `g:Abolish.snakecase()`.

Reply:
Thanks for the example, the patching indeed works and calling `g:Abolish.snakecase()` works, too.


> Doesn't sound like you need to. Override the `_` coercion, or whatever.

This only works like I did it above though, right? Because I cannot overwrite the `snakecase` function and call it internally like for example so:
```vim
function! s:snakecase(word)
  return substitute(g:Abolish.snakecase(a:word),'/','_','g')
endfunction

call extend(Abolish, {'snakecase': s:function('s:snakecase')}, "keep")

```
I was hoping that something like this could be done because in this way I would continue receiving updates that our done upstream. But this leads to an endless recursion.

And while overwriting the `_` `Abolish.Coercions` does make the `/ -> _` reversal work, this does not integrate it into the other coercions becacuse those have the `s:snakecase` or `Abolish.snakecase` function hard-coded. Or am I overlooking something here?

Reply:
> Thanks for the example, the patching indeed works and calling `g:Abolish.snakecase()` works, too.
> 
> > Doesn't sound like you need to. Override the `_` coercion, or whatever.
> 
> This only works like I did it above though, right? Because I cannot overwrite the `snakecase` function and call it internally like for example so:
> 
> ```viml
> function! s:snakecase(word)
>   return substitute(g:Abolish.snakecase(a:word),'/','_','g')
> endfunction
> 
> call extend(Abolish, {'snakecase': s:function('s:snakecase')}, "keep")
> ```
> 
> I was hoping that something like this could be done because in this way I would continue receiving updates that our done upstream. But this leads to an endless recursion.

You can avoid recursion by sticking it in a variable first, and referencing that:

```vim
let s:abolish_snakecase = g:Abolish.snakecase
```

> And while overwriting the `_` `Abolish.Coercions` does make the `/ -> _` reversal work, this does not integrate it into the other coercions becacuse those have the `s:snakecase` or `Abolish.snakecase` function hard-coded. Or am I overlooking something here?

That hard-coding could certainly be changed, but it kind of sounds like you're preemptively overgeneralizing? Based on your original use case, it seems like all you need is `.` to `/` and vice versa. Perhaps replace the `.` coercion, add a `/`, and call it day.

Reply:
Okay thanks for the tip! Indeed, I guess only `.` and `/` would be sufficient.

However, one final question: would you be willing to accept a PR which changes `s:snakecase` to `g:Abolish.snakecase`? That would fix all edge cases and allow the exact monkey-patching that I am looking for to make it work for all case :slightly_smiling_face: 

If not that is fine, too. In that case we can consider this (and the matching issue) closed :+1:

Reply:
> However, one final question: would you be willing to accept a PR which changes `s:snakecase` to `g:Abolish.snakecase`? That would fix all edge cases and allow the exact monkey-patching that I am looking for to make it work for all case

I have but one objection to this and that's that it's unclear what _exactly_ should change. You want to change the internal implementations of coercions that delegate to it, obviously, but what about the `_` and `s` coercions themselves, which [hard-code a reference to it globally when the plugin is loaded](https://github.com/tpope/vim-rails/blob/c3ac2889dc448c5442fa2d1096ac918ea8b38f09/plugin/rails.vim#L152-L183) and thus wouldn't be impacted by a retroactive monkey-patch? What about other helpers like `s:camelcase()`, should those be changed too? What about non-coercion functionality, should that also be changed? A monkey-patch affecting `:Subvert` worries me a lot more than one just affecting coercions, but is it worth the inconsistency to special case that?

> I was hoping that something like this could be done because in this way I would continue receiving updates that our done upstream.

These functions literally never change. This benefit is purely hypothetical. I'm gonna needs something more substantial if I'm going to open this can of worms.